core/verkle_witness_test.go

```diff
@@ -0,0 +1,999 @@
+		gen.SetPoS()
+		if i == 0 {
+			// Create self-destruct contract, sending 42 wei.
+			tx, _ := types.SignTx(types.NewContractCreation(0, big.NewInt(42), 100_000, big.NewInt(875000000), selfDestructContract), signer, testKey)
+			gen.AddTx(tx)
+		} else {
+			// Call it.
+			tx, _ := types.SignTx(types.NewTransaction(1, contract, big.NewInt(0), 100_000, big.NewInt(875000000), nil), signer, testKey)
+			gen.AddTx(tx)
+		}
+	})
+
+	{
+		// Check self-destructed contract in the witness.
+		// The way 6780 is implemented today, it always SubBalance from the self-destructed contract, and AddBalance
+		// to the beneficiary. In this case both addresses are the same, thus this might be optimizable from a gas
+		// perspective. But until that happens, we need to honor this "balance reading" adding it to the witness.
+
+		selfDestructContractTreeKey := utils.CodeHashKey(contract[:])
+
+		var stateDiffIdx = -1
+		for i, stemStateDiff := range statediffs[1] {
+			if bytes.Equal(stemStateDiff.Stem[:], selfDestructContractTreeKey[:31]) {
+				stateDiffIdx = i
+				break
+			}
+		}
+		if stateDiffIdx == -1 {
+			t.Fatal("no state diff found for stem")
+		}
+
+		balanceStateDiff := statediffs[1][stateDiffIdx].SuffixDiffs[0]
+		if balanceStateDiff.Suffix != utils.BasicDataLeafKey {
+			t.Fatal("balance invalid suffix")
+		}
+
+		// The original balance was 42.
+		var oldBalance [16]byte
+		oldBalance[15] = 42
+		if !bytes.Equal((*balanceStateDiff.CurrentValue)[utils.BasicDataBalanceOffset:], oldBalance[:]) {
+			t.Fatal("the pre-state balance before self-destruct must be 42")
+		}
+
+		// Note that the SubBalance+AddBalance net effect is a 0 change, so NewValue
+		// must be nil.
+		if balanceStateDiff.NewValue != nil {
+			t.Fatal("the post-state balance after self-destruct must be empty")
+		}
+	}
+}
+
+// TestProcessVerkleSelfDestructInSameTxWithSelfBeneficiary checks the content of the witness
+// if a selfdestruct occurs in the same tx as the one that created it, but the beneficiary
+// is the selfdestructed account.
+func TestProcessVerkleSelfDestructInSameTxWithSelfBeneficiary(t *testing.T) {
+	// The test txs were taken from a secondary testnet with chain id 69421
+	config := *testKaustinenLikeChainConfig
+	config.ChainID.SetUint64(69421)
+
+	var (
+		signer     = types.LatestSigner(&config)
+		testKey, _ = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
+		gspec      = verkleTestGenesis(&config)
+		deployer   = crypto.PubkeyToAddress(testKey.PublicKey)
+		contract   = crypto.CreateAddress(deployer, 0)
+	)
+	// Also add the contract-to-destroy
+	gspec.Alloc[contract] = &types.Account{
+		Code:       nil,
+		Storage:    nil,
+		Balance:    nil,
+		Nonce:      0,
+		PrivateKey: nil,
+	}
+	// The goal of this test is to test SELFDESTRUCT that happens in a contract
+	// execution which is created in **the same** transaction sending the remaining
+	// balance to itself.
```

Comment:
This test doesn't achieve that goal. The runtime bytecode is correct, it does a selfdestruct-to-self. However, the test passes even if one uses e.g. `selfDestructContract := []byte{byte(vm.NUMBER), byte(vm.SELFDESTRUCT)}`

But, if I add some money to it beforehand, 
```
	// Also add the contract-to-destroy
	gspec.Alloc[contract] = types.Account{
		Balance: big.NewInt(100),
	}
```
then we get a crash: 
```
=== RUN   TestProcessVerkleSelfDestructInSameTxWithSelfBeneficiary
    verkle_witness_test.go:958: Contract: 0x3A220f351252089D385b29beca14e27F204c296A
--- FAIL: TestProcessVerkleSelfDestructInSameTxWithSelfBeneficiary (1.51s)
panic: not implemented [recovered]
	panic: not implemented

goroutine 29 [running]:
testing.tRunner.func1.2({0xebcb80, 0x1290420})
	/usr/local/go/src/testing/testing.go:1632 +0x230
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1635 +0x35e
panic({0xebcb80?, 0x1290420?})
	/usr/local/go/src/runtime/panic.go:785 +0x132
github.com/ethereum/go-ethereum/trie.(*VerkleTrie).NodeIterator(0xa4bd96820128063b?, {0x6539b885bc236522?, 0x1f81a32f5e61a48c?, 0xaceb1d6fa1cb7083?})
	/home/user/go/src/github.com/ethereum/go-ethereum/trie/verkle.go:255 +0x25
```

I'll commit the change which makes it break, so we don't forget to fix it :)

Reply:
> This test doesn't achieve that goal. The runtime bytecode is correct, it does a selfdestruct-to-self. However, the test passes even if one uses e.g. `selfDestructContract := []byte{byte(vm.NUMBER), byte(vm.SELFDESTRUCT)}`

It does, the bug I'm interested in, which was fixed, is that if you self-destruct with yourself as a beneficiary, the funds don't end up in the witness - it should be nothing. If you selfdestruct to another address, the balance isn't transferred to the self address, so the "new value" will still be `nil`, which is normal, and the test passes.

> 
> But, if I add some money to it beforehand,

ETH is not money :grin: 

> 
> ```
> 	// Also add the contract-to-destroy
> 	gspec.Alloc[contract] = types.Account{
> 		Balance: big.NewInt(100),
> 	}
> ```

This means you are no longer testing the use case, as the contract already exists - which is still a valid test case, but you're changing what is being tested.

> 
> then we get a crash:
> 
> ```
> === RUN   TestProcessVerkleSelfDestructInSameTxWithSelfBeneficiary
>     verkle_witness_test.go:958: Contract: 0x3A220f351252089D385b29beca14e27F204c296A
> --- FAIL: TestProcessVerkleSelfDestructInSameTxWithSelfBeneficiary (1.51s)
> panic: not implemented [recovered]
> 	panic: not implemented
> 
> goroutine 29 [running]:
> testing.tRunner.func1.2({0xebcb80, 0x1290420})
> 	/usr/local/go/src/testing/testing.go:1632 +0x230
> testing.tRunner.func1()
> 	/usr/local/go/src/testing/testing.go:1635 +0x35e
> panic({0xebcb80?, 0x1290420?})
> 	/usr/local/go/src/runtime/panic.go:785 +0x132
> github.com/ethereum/go-ethereum/trie.(*VerkleTrie).NodeIterator(0xa4bd96820128063b?, {0x6539b885bc236522?, 0x1f81a32f5e61a48c?, 0xaceb1d6fa1cb7083?})
> 	/home/user/go/src/github.com/ethereum/go-ethereum/trie/verkle.go:255 +0x25
> ```
> 
> I'll commit the change which makes it break, so we don't forget to fix it :)

Ok yeah, interestingly that triggered a piece of code that shouldn't be called. I have to look into it, thanks.